src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs

```diff
@@ -268,59 +265,74 @@ public JsonClassInfo(Type type, JsonSerializerOptions options)
                 {
-                    // This is not null because it is an actual
-                    // property on a type, not a "policy property".
-                    MemberInfo memberInfo = jsonPropertyInfo.MemberInfo!;
-                    Debug.Assert(memberInfo is PropertyInfo || memberInfo is FieldInfo);
-
-                    Type memberType = memberInfo is PropertyInfo propertyInfo
-                        ? propertyInfo.PropertyType
-                        : Unsafe.As<FieldInfo>(memberInfo).FieldType;
+                    // More than one property has the same case-insensitive name.
+                    // Remember so we can throw a nice exception if this property is used as a parameter name.
+                    nameLookup[propertyName].DuplicateJsonPropertyInfo = jsonProperty;
+                }
+            }
 
-                    string camelCasePropName = JsonNamingPolicy.CamelCase.ConvertName(memberInfo.Name);
+            Type GetMemberType(MemberInfo memberInfo)
+            {
+                Debug.Assert(memberInfo is PropertyInfo || memberInfo is FieldInfo);
 
-                    if (parameterInfo.Name == camelCasePropName &&
-                        parameterInfo.ParameterType == memberType)
-                    {
-                        if (isBound)
-                        {
-                            Debug.Assert(firstMatch != null);
-
-                            // Multiple object properties cannot bind to the same
-                            // constructor parameter.
-                            ThrowHelper.ThrowInvalidOperationException_MultiplePropertiesBindToConstructorParameters(
-                                Type,
-                                parameterInfo,
-                                firstMatch,
-                                memberInfo,
-                                constructorInfo);
-                        }
+                return memberInfo is PropertyInfo propertyInfo
+                    ? propertyInfo.PropertyType
+                    : Unsafe.As<FieldInfo>(memberInfo).FieldType;
+            }
 
-                        JsonParameterInfo jsonParameterInfo = AddConstructorParameter(parameterInfo, jsonPropertyInfo, Options);
+            foreach (ParameterInfo parameterInfo in parameters)
+            {
+                string parameterName = parameterInfo.Name!;
 
-                        // One object property cannot map to multiple constructor
-                        // parameters (ConvertName above can't return multiple strings).
-                        parameterCache.Add(jsonPropertyInfo.NameAsString, jsonParameterInfo);
+                if (nameLookup.TryGetValue(parameterName, out ParameterLookupEntry? matchingEntry) &&
+                    (parameterInfo.ParameterType == GetMemberType(matchingEntry.JsonPropertyInfo.MemberInfo!)))
+                {
+                    if (matchingEntry.DuplicateJsonPropertyInfo != null)
```

Comment:
Do we need to check if `DuplicateJsonPropertyInfo` has the same member type? Consider

```C#
public record AgeRecord(int age)
{
   public string Age { get; set; } = age.ToString();
}
```

Reply:
Yes thanks. I updated the PR to support this scenario. This was an untested case (both for POCOs and records) so I added tests.